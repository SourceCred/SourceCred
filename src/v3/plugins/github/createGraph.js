// @flow

import {Graph} from "../../core/graph";
import type {
  GithubResponseJSON,
  RepositoryJSON,
  ReviewJSON,
  PullJSON,
  IssueJSON,
  CommentJSON,
  ReviewCommentJSON,
  NullableAuthorJSON,
} from "./graphql";

import type {
  RepoAddress,
  IssueAddress,
  PullAddress,
  ReviewAddress,
  UserlikeAddress,
  StructuredAddress,
  AuthorableAddress,
  ChildAddress,
  ParentAddress,
} from "./nodes";
import {toRaw} from "./nodes";

import {createEdge} from "./edges";

import {COMMIT_TYPE, toRaw as gitToRaw} from "../git/nodes";

import {
  reviewUrlToId,
  issueCommentUrlToId,
  pullCommentUrlToId,
  reviewCommentUrlToId,
} from "./urlIdParse";

export function createGraph(data: GithubResponseJSON): Graph {
  const creator = new GraphCreator();
  creator.addData(data);
  return creator.graph;
}

class GraphCreator {
  graph: Graph;

  constructor() {
    this.graph = new Graph();
  }

  addNode(addr: StructuredAddress) {
    this.graph.addNode(toRaw(addr));
  }

  addData(data: GithubResponseJSON) {
    this.addRepository(data.repository);
  }

  addRepository(repoJSON: RepositoryJSON) {
    const repo: RepoAddress = {
      type: "REPO",
      owner: repoJSON.owner.login,
      name: repoJSON.name,
    };
    this.addNode(repo);
    repoJSON.issues.nodes.forEach((issue) => this.addIssue(repo, issue));
    repoJSON.pulls.nodes.forEach((pull) => this.addPull(repo, pull));
  }

  addIssue(repo: RepoAddress, issueJSON: IssueJSON) {
    const issue: IssueAddress = {
      type: "ISSUE",
      repo,
      number: String(issueJSON.number),
    };
    this.addNode(issue);
    this.addAuthors(issue, issueJSON.author);
    this.addHasParent(issue, repo);
    issueJSON.comments.nodes.forEach((comment) =>
      this.addComment(issue, comment)
    );
  }

  addPull(repo: RepoAddress, pullJSON: PullJSON) {
    const pull: PullAddress = {
      type: "PULL",
      repo,
      number: String(pullJSON.number),
    };
    this.addNode(pull);
    this.addAuthors(pull, pullJSON.author);
    this.addHasParent(pull, repo);
    pullJSON.comments.nodes.forEach((c) => this.addComment(pull, c));
    pullJSON.reviews.nodes.forEach((review) => this.addReview(pull, review));
    if (pullJSON.mergeCommit != null) {
      const commitHash = pullJSON.mergeCommit.oid;
      const commit = {type: COMMIT_TYPE, hash: commitHash};
      this.graph.addNode(gitToRaw(commit));
      this.graph.addEdge(createEdge.mergedAs(pull, commit));
    }
  }

  addReview(pull: PullAddress, reviewJSON: ReviewJSON) {
    const id = reviewUrlToId(reviewJSON.url);
    const review = {
      type: "REVIEW",
      pull,
      id,
    };
    this.addNode(review);
    reviewJSON.comments.nodes.forEach((c) => this.addComment(review, c));
    this.addAuthors(review, reviewJSON.author);
    this.addHasParent(review, pull);
  }

  addComment(
    parent: IssueAddress | PullAddress | ReviewAddress,
    commentJSON: CommentJSON | ReviewCommentJSON
  ) {
    const id = (function() {
      switch (parent.type) {
        case "ISSUE":
          return issueCommentUrlToId(commentJSON.url);
        case "PULL":
          return pullCommentUrlToId(commentJSON.url);
        case "REVIEW":
          return reviewCommentUrlToId(commentJSON.url);
        default:
          // eslint-disable-next-line no-unused-expressions
          (parent.type: empty);
          throw new Error(`Unexpected comment parent type: ${parent.type}`);
      }
    })();
    const comment = {
      type: "COMMENT",
      parent,
      id,
    };
    this.addNode(comment);
    this.addAuthors(comment, commentJSON.author);
    this.addHasParent(comment, parent);
  }

  addAuthors(content: AuthorableAddress, authorJSON: NullableAuthorJSON) {
    // author may be null, as not all posts have authors
    if (authorJSON == null) {
      return;
    }
    const author: UserlikeAddress = {
      type: "USERLIKE",
      login: authorJSON.login,
    };
    this.addNode(author);
    this.graph.addEdge(createEdge.authors(author, content));
  }

  addHasParent(child: ChildAddress, parent: ParentAddress) {
    this.graph.addEdge(createEdge.hasParent(child, parent));
  }
}
