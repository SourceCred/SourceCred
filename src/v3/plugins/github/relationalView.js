// @flow

import * as N from "./nodes";
import * as Q from "./graphql";
import * as GitNode from "../git/nodes";

import {
  reviewUrlToId,
  issueCommentUrlToId,
  pullCommentUrlToId,
  reviewCommentUrlToId,
} from "./urlIdParse";

export type RepoEntry = {|
  +type: typeof N.REPO_TYPE,
  +address: N.RepoAddress,
  +url: string,
  +issues: IssueEntry[],
  +pulls: PullEntry[],
|};

export type IssueEntry = {|
  +type: typeof N.ISSUE_TYPE,
  +address: N.IssueAddress,
  +title: string,
  +body: string,
  +url: string,
  +comments: CommentEntry[],
  +nominalAuthor: ?UserlikeEntry,
|};

export type PullEntry = {|
  +type: typeof N.PULL_TYPE,
  +address: N.PullAddress,
  +title: string,
  +body: string,
  +url: string,
  +comments: CommentEntry[],
  +reviews: ReviewEntry[],
  +mergedAs: ?GitNode.CommitAddress,
  +nominalAuthor: ?UserlikeEntry,
|};

export type ReviewEntry = {|
  +type: typeof N.REVIEW_TYPE,
  +address: N.ReviewAddress,
  +body: string,
  +url: string,
  +comments: CommentEntry[],
  +state: Q.ReviewState,
  +nominalAuthor: ?UserlikeEntry,
|};

export type CommentEntry = {|
  +type: typeof N.COMMENT_TYPE,
  +address: N.CommentAddress,
  +body: string,
  +url: string,
  +nominalAuthor: ?UserlikeEntry,
|};

export type UserlikeEntry = {|
  +type: typeof N.USERLIKE_TYPE,
  +address: N.UserlikeAddress,
  +url: string,
|};

export type Entry =
  | RepoEntry
  | IssueEntry
  | PullEntry
  | ReviewEntry
  | CommentEntry
  | UserlikeEntry;

export class RelationalView {
  _repos: Map<N.RawAddress, RepoEntry>;
  _issues: Map<N.RawAddress, IssueEntry>;
  _pulls: Map<N.RawAddress, PullEntry>;
  _comments: Map<N.RawAddress, CommentEntry>;
  _reviews: Map<N.RawAddress, ReviewEntry>;
  _userlikes: Map<N.RawAddress, UserlikeEntry>;

  constructor(data: Q.GithubResponseJSON) {
    this._repos = new Map();
    this._issues = new Map();
    this._pulls = new Map();
    this._comments = new Map();
    this._reviews = new Map();
    this._userlikes = new Map();
    this._addRepo(data.repository);
  }

  repos(): Iterator<RepoEntry> {
    return this._repos.values();
  }
  repo(address: N.RepoAddress): ?RepoEntry {
    return this._repos.get(N.toRaw(address));
  }

  issues(): Iterator<IssueEntry> {
    return this._issues.values();
  }
  issue(address: N.IssueAddress): ?IssueEntry {
    return this._issues.get(N.toRaw(address));
  }

  pulls(): Iterator<PullEntry> {
    return this._pulls.values();
  }
  pull(address: N.PullAddress): ?PullEntry {
    return this._pulls.get(N.toRaw(address));
  }

  comments(): Iterator<CommentEntry> {
    return this._comments.values();
  }
  comment(address: N.CommentAddress): ?CommentEntry {
    return this._comments.get(N.toRaw(address));
  }

  reviews(): Iterator<ReviewEntry> {
    return this._reviews.values();
  }
  review(address: N.ReviewAddress): ?ReviewEntry {
    return this._reviews.get(N.toRaw(address));
  }

  userlikes(): Iterator<UserlikeEntry> {
    return this._userlikes.values();
  }
  userlike(address: N.UserlikeAddress): ?UserlikeEntry {
    return this._userlikes.get(N.toRaw(address));
  }

  _addRepo(json: Q.RepositoryJSON) {
    const address: N.RepoAddress = {
      type: N.REPO_TYPE,
      owner: json.owner.login,
      name: json.name,
    };
    const entry: RepoEntry = {
      address,
      url: json.url,
      issues: json.issues.nodes.map((x) => this._addIssue(address, x)),
      pulls: json.pulls.nodes.map((x) => this._addPull(address, x)),
      type: N.REPO_TYPE,
    };
    const raw = N.toRaw(address);
    this._repos.set(raw, entry);
  }

  _addIssue(repo: N.RepoAddress, json: Q.IssueJSON): IssueEntry {
    const address: N.IssueAddress = {
      type: N.ISSUE_TYPE,
      number: String(json.number),
      repo,
    };
    const entry: IssueEntry = {
      address,
      url: json.url,
      comments: json.comments.nodes.map((x) => this._addComment(address, x)),
      nominalAuthor: this._addNullableAuthor(json.author),
      body: json.body,
      title: json.title,
      type: N.ISSUE_TYPE,
    };
    this._issues.set(N.toRaw(address), entry);
    return entry;
  }

  _addPull(repo: N.RepoAddress, json: Q.PullJSON): PullEntry {
    const address: N.PullAddress = {
      type: N.PULL_TYPE,
      number: String(json.number),
      repo,
    };
    const mergedAs =
      json.mergeCommit == null
        ? null
        : {
            type: GitNode.COMMIT_TYPE,
            hash: json.mergeCommit.oid,
          };

    const entry: PullEntry = {
      address,
      url: json.url,
      comments: json.comments.nodes.map((x) => this._addComment(address, x)),
      reviews: json.reviews.nodes.map((x) => this._addReview(address, x)),
      nominalAuthor: this._addNullableAuthor(json.author),
      body: json.body,
      title: json.title,
      mergedAs,
      type: N.PULL_TYPE,
    };
    this._pulls.set(N.toRaw(address), entry);
    return entry;
  }

  _addReview(pull: N.PullAddress, json: Q.ReviewJSON): ReviewEntry {
    const address: N.ReviewAddress = {
      type: N.REVIEW_TYPE,
      id: reviewUrlToId(json.url),
      pull,
    };
    const entry: ReviewEntry = {
      address,
      url: json.url,
      state: json.state,
      comments: json.comments.nodes.map((x) => this._addComment(address, x)),
      body: json.body,
      nominalAuthor: this._addNullableAuthor(json.author),
      type: N.REVIEW_TYPE,
    };
    this._reviews.set(N.toRaw(address), entry);
    return entry;
  }

  _addComment(
    parent: N.IssueAddress | N.PullAddress | N.ReviewAddress,
    json: Q.CommentJSON
  ): CommentEntry {
    const id = (function() {
      switch (parent.type) {
        case N.ISSUE_TYPE:
          return issueCommentUrlToId(json.url);
        case N.PULL_TYPE:
          return pullCommentUrlToId(json.url);
        case N.REVIEW_TYPE:
          return reviewCommentUrlToId(json.url);
        default:
          // eslint-disable-next-line no-unused-expressions
          (parent.type: empty);
          throw new Error(`Unexpected comment parent type: ${parent.type}`);
      }
    })();
    const address: N.CommentAddress = {type: N.COMMENT_TYPE, id, parent};
    const entry: CommentEntry = {
      address,
      url: json.url,
      nominalAuthor: this._addNullableAuthor(json.author),
      body: json.body,
      type: N.COMMENT_TYPE,
    };
    this._comments.set(N.toRaw(address), entry);
    return entry;
  }

  _addNullableAuthor(json: Q.NullableAuthorJSON): ?UserlikeEntry {
    if (json == null) {
      return null;
    } else {
      const address: N.UserlikeAddress = {
        type: N.USERLIKE_TYPE,
        login: json.login,
      };
      const entry: UserlikeEntry = {
        address,
        url: json.url,
        type: N.USERLIKE_TYPE,
      };
      this._userlikes.set(N.toRaw(address), entry);
      return entry;
    }
  }
}
