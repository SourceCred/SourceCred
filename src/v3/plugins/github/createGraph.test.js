// @flow

import {createGraph} from "./createGraph";
import {GraphView} from "./graphView";
import {RelationalView} from "./relationalView";
import type {GithubResponseJSON} from "./graphql";
import cloneDeep from "lodash.clonedeep";

function exampleGraph() {
  const data: GithubResponseJSON = cloneDeep(
    require("./demoData/example-github")
  );
  const view = new RelationalView(data);
  return createGraph(view);
}

describe("plugins/github/createGraph", () => {
  it("example graph matches snapshot", () => {
    expect(exampleGraph()).toMatchSnapshot();
  });

  it("passes all GraphView invariants", () => {
    const graph = exampleGraph();
    const view = new GraphView(graph);
    // This test is high leverage. It checks:
    // - that every node starting with a GitHub prefix
    //   - can be structured using fromRaw
    //   - has the correct type
    // - that every edge starting with a GitHub prefix
    //   - can be structured using fromRaw
    //   - and has the correct type,
    //   - and that its src has an expected prefix,
    //   - and that its dst has an expected prefix,
    // - that every child node
    //   - has exactly one parent
    //   - has a parent with the correct type
    view.checkInvariants();
    // as currently written, GV checks invariants on construction.
    // we call the method explicitly as a defensive step.
  });
});
