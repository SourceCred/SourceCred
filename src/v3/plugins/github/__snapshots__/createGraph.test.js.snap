// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/github/createGraph example graph matches snapshot 1`] = `
Array [
  Object {
    "type": "sourcecred/graph",
    "version": "0.4.0",
  },
  Object {
    "edges": Array [
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "4",
          "ISSUE",
          "sourcecred",
          "example-github",
          "1",
        ],
        "dstIndex": 16,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "4",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
        ],
        "dstIndex": 17,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "4",
          "ISSUE",
          "sourcecred",
          "example-github",
          "4",
        ],
        "dstIndex": 18,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "4",
          "ISSUE",
          "sourcecred",
          "example-github",
          "6",
        ],
        "dstIndex": 19,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "4",
          "ISSUE",
          "sourcecred",
          "example-github",
          "7",
        ],
        "dstIndex": 20,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "4",
          "ISSUE",
          "sourcecred",
          "example-github",
          "8",
        ],
        "dstIndex": 21,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "4",
          "PULL",
          "sourcecred",
          "example-github",
          "3",
        ],
        "dstIndex": 22,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "4",
          "PULL",
          "sourcecred",
          "example-github",
          "5",
        ],
        "dstIndex": 23,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "4",
          "PULL",
          "sourcecred",
          "example-github",
          "9",
        ],
        "dstIndex": 24,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
          "373768703",
        ],
        "dstIndex": 2,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
          "373768850",
        ],
        "dstIndex": 3,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
          "385576185",
        ],
        "dstIndex": 4,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
          "385576220",
        ],
        "dstIndex": 5,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
          "385576248",
        ],
        "dstIndex": 6,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
          "385576273",
        ],
        "dstIndex": 7,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
          "385576920",
        ],
        "dstIndex": 8,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
          "385576936",
        ],
        "dstIndex": 9,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "6",
          "373768442",
        ],
        "dstIndex": 10,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "6",
          "373768538",
        ],
        "dstIndex": 11,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "6",
          "385223316",
        ],
        "dstIndex": 12,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "decentralion",
          "6",
          "COMMENT",
          "PULL",
          "sourcecred",
          "example-github",
          "3",
          "369162222",
        ],
        "dstIndex": 13,
        "srcIndex": 28,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "wchargin",
          "5",
          "REVIEW",
          "sourcecred",
          "example-github",
          "5",
          "100313899",
        ],
        "dstIndex": 26,
        "srcIndex": 29,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "wchargin",
          "5",
          "REVIEW",
          "sourcecred",
          "example-github",
          "5",
          "100314038",
        ],
        "dstIndex": 27,
        "srcIndex": 29,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "wchargin",
          "6",
          "COMMENT",
          "PULL",
          "sourcecred",
          "example-github",
          "5",
          "396430464",
        ],
        "dstIndex": 14,
        "srcIndex": 29,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "AUTHORS",
          "2",
          "USERLIKE",
          "wchargin",
          "7",
          "COMMENT",
          "REVIEW",
          "sourcecred",
          "example-github",
          "5",
          "100313899",
          "171460198",
        ],
        "dstIndex": 15,
        "srcIndex": 29,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "4",
          "ISSUE",
          "sourcecred",
          "example-github",
          "1",
        ],
        "dstIndex": 25,
        "srcIndex": 16,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "4",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
        ],
        "dstIndex": 25,
        "srcIndex": 17,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "4",
          "ISSUE",
          "sourcecred",
          "example-github",
          "4",
        ],
        "dstIndex": 25,
        "srcIndex": 18,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "4",
          "ISSUE",
          "sourcecred",
          "example-github",
          "6",
        ],
        "dstIndex": 25,
        "srcIndex": 19,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "4",
          "ISSUE",
          "sourcecred",
          "example-github",
          "7",
        ],
        "dstIndex": 25,
        "srcIndex": 20,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "4",
          "ISSUE",
          "sourcecred",
          "example-github",
          "8",
        ],
        "dstIndex": 25,
        "srcIndex": 21,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "4",
          "PULL",
          "sourcecred",
          "example-github",
          "3",
        ],
        "dstIndex": 25,
        "srcIndex": 22,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "4",
          "PULL",
          "sourcecred",
          "example-github",
          "5",
        ],
        "dstIndex": 25,
        "srcIndex": 23,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "4",
          "PULL",
          "sourcecred",
          "example-github",
          "9",
        ],
        "dstIndex": 25,
        "srcIndex": 24,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "5",
          "REVIEW",
          "sourcecred",
          "example-github",
          "5",
          "100313899",
        ],
        "dstIndex": 23,
        "srcIndex": 26,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "5",
          "REVIEW",
          "sourcecred",
          "example-github",
          "5",
          "100314038",
        ],
        "dstIndex": 23,
        "srcIndex": 27,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
          "373768703",
        ],
        "dstIndex": 17,
        "srcIndex": 2,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
          "373768850",
        ],
        "dstIndex": 17,
        "srcIndex": 3,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
          "385576185",
        ],
        "dstIndex": 17,
        "srcIndex": 4,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
          "385576220",
        ],
        "dstIndex": 17,
        "srcIndex": 5,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
          "385576248",
        ],
        "dstIndex": 17,
        "srcIndex": 6,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
          "385576273",
        ],
        "dstIndex": 17,
        "srcIndex": 7,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
          "385576920",
        ],
        "dstIndex": 17,
        "srcIndex": 8,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "2",
          "385576936",
        ],
        "dstIndex": 17,
        "srcIndex": 9,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "6",
          "373768442",
        ],
        "dstIndex": 19,
        "srcIndex": 10,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "6",
          "373768538",
        ],
        "dstIndex": 19,
        "srcIndex": 11,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "6",
          "COMMENT",
          "ISSUE",
          "sourcecred",
          "example-github",
          "6",
          "385223316",
        ],
        "dstIndex": 19,
        "srcIndex": 12,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "6",
          "COMMENT",
          "PULL",
          "sourcecred",
          "example-github",
          "3",
          "369162222",
        ],
        "dstIndex": 22,
        "srcIndex": 13,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "6",
          "COMMENT",
          "PULL",
          "sourcecred",
          "example-github",
          "5",
          "396430464",
        ],
        "dstIndex": 23,
        "srcIndex": 14,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "HAS_PARENT",
          "7",
          "COMMENT",
          "REVIEW",
          "sourcecred",
          "example-github",
          "5",
          "100313899",
          "171460198",
        ],
        "dstIndex": 26,
        "srcIndex": 15,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "MERGED_AS",
          "4",
          "PULL",
          "sourcecred",
          "example-github",
          "3",
        ],
        "dstIndex": 0,
        "srcIndex": 22,
      },
      Object {
        "address": Array [
          "sourcecred",
          "github",
          "MERGED_AS",
          "4",
          "PULL",
          "sourcecred",
          "example-github",
          "5",
        ],
        "dstIndex": 1,
        "srcIndex": 23,
      },
    ],
    "nodes": Array [
      Array [
        "sourcecred",
        "git",
        "COMMIT",
        "0a223346b4e6dec0127b1e6aa892c4ee0424b66a",
      ],
      Array [
        "sourcecred",
        "git",
        "COMMIT",
        "6d5b3aa31ebb68a06ceb46bbd6cf49b6ccd6f5e6",
      ],
      Array [
        "sourcecred",
        "github",
        "COMMENT",
        "ISSUE",
        "sourcecred",
        "example-github",
        "2",
        "373768703",
      ],
      Array [
        "sourcecred",
        "github",
        "COMMENT",
        "ISSUE",
        "sourcecred",
        "example-github",
        "2",
        "373768850",
      ],
      Array [
        "sourcecred",
        "github",
        "COMMENT",
        "ISSUE",
        "sourcecred",
        "example-github",
        "2",
        "385576185",
      ],
      Array [
        "sourcecred",
        "github",
        "COMMENT",
        "ISSUE",
        "sourcecred",
        "example-github",
        "2",
        "385576220",
      ],
      Array [
        "sourcecred",
        "github",
        "COMMENT",
        "ISSUE",
        "sourcecred",
        "example-github",
        "2",
        "385576248",
      ],
      Array [
        "sourcecred",
        "github",
        "COMMENT",
        "ISSUE",
        "sourcecred",
        "example-github",
        "2",
        "385576273",
      ],
      Array [
        "sourcecred",
        "github",
        "COMMENT",
        "ISSUE",
        "sourcecred",
        "example-github",
        "2",
        "385576920",
      ],
      Array [
        "sourcecred",
        "github",
        "COMMENT",
        "ISSUE",
        "sourcecred",
        "example-github",
        "2",
        "385576936",
      ],
      Array [
        "sourcecred",
        "github",
        "COMMENT",
        "ISSUE",
        "sourcecred",
        "example-github",
        "6",
        "373768442",
      ],
      Array [
        "sourcecred",
        "github",
        "COMMENT",
        "ISSUE",
        "sourcecred",
        "example-github",
        "6",
        "373768538",
      ],
      Array [
        "sourcecred",
        "github",
        "COMMENT",
        "ISSUE",
        "sourcecred",
        "example-github",
        "6",
        "385223316",
      ],
      Array [
        "sourcecred",
        "github",
        "COMMENT",
        "PULL",
        "sourcecred",
        "example-github",
        "3",
        "369162222",
      ],
      Array [
        "sourcecred",
        "github",
        "COMMENT",
        "PULL",
        "sourcecred",
        "example-github",
        "5",
        "396430464",
      ],
      Array [
        "sourcecred",
        "github",
        "COMMENT",
        "REVIEW",
        "sourcecred",
        "example-github",
        "5",
        "100313899",
        "171460198",
      ],
      Array [
        "sourcecred",
        "github",
        "ISSUE",
        "sourcecred",
        "example-github",
        "1",
      ],
      Array [
        "sourcecred",
        "github",
        "ISSUE",
        "sourcecred",
        "example-github",
        "2",
      ],
      Array [
        "sourcecred",
        "github",
        "ISSUE",
        "sourcecred",
        "example-github",
        "4",
      ],
      Array [
        "sourcecred",
        "github",
        "ISSUE",
        "sourcecred",
        "example-github",
        "6",
      ],
      Array [
        "sourcecred",
        "github",
        "ISSUE",
        "sourcecred",
        "example-github",
        "7",
      ],
      Array [
        "sourcecred",
        "github",
        "ISSUE",
        "sourcecred",
        "example-github",
        "8",
      ],
      Array [
        "sourcecred",
        "github",
        "PULL",
        "sourcecred",
        "example-github",
        "3",
      ],
      Array [
        "sourcecred",
        "github",
        "PULL",
        "sourcecred",
        "example-github",
        "5",
      ],
      Array [
        "sourcecred",
        "github",
        "PULL",
        "sourcecred",
        "example-github",
        "9",
      ],
      Array [
        "sourcecred",
        "github",
        "REPO",
        "sourcecred",
        "example-github",
      ],
      Array [
        "sourcecred",
        "github",
        "REVIEW",
        "sourcecred",
        "example-github",
        "5",
        "100313899",
      ],
      Array [
        "sourcecred",
        "github",
        "REVIEW",
        "sourcecred",
        "example-github",
        "5",
        "100314038",
      ],
      Array [
        "sourcecred",
        "github",
        "USERLIKE",
        "decentralion",
      ],
      Array [
        "sourcecred",
        "github",
        "USERLIKE",
        "wchargin",
      ],
    ],
  },
]
`;
